### PR TITLE
Data statistics manager

### PR DESCRIPTION
**Pull request by datomo**
This pull request introduces the underlying system to enable dynamic querying. 
This system automaticly suggests refinement options when querying with the Graphical Querying.
Two different ways of tracking statistics is introduced, which, like the whole extension, can be configured through the _ConfigManager_.

**Major Changes:**
- _StatisticsManager_ handles the collecting and storing of the needed data.
- _SQLStatisticProcessor_ enables sending SQL queries from the _StatisticsManager_

**Added Configs**
- useDynamicQuerying: can disable the tracking of statistics and the manager itself
- activeTracking: toggles the tracking of dml operations through the frontend
- passiveTracking: toggles the refreshing of  all statistics by a set rate, disabled by default
- StatisticColumnBuffer: amount of saved values for e.g. unique values
- maxCharUniqueVal: maximum of characters of unique values

### CODE REVIEW

**Comment by vogti** [comment]
@schoenja can you please take a look at this PR as well?

**Comment by datomo** [comment]
I realised that the applying of specific statistic after commiting a transaction and the whole reevaluating of the statisticManager should run in seperate threads. I'll change that as well.

- [ ] reevaluating of whole manager in thread
- [ ] applying in own thread

File: core/src/main/java/ch/unibas/dmi/dbis/polyphenydb/statistic/StatisticQueryProcessor.java [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+
+                types.add( PolySqlType.getPolySqlTypeFromSting( metaData.type.name ) );
+                names.add( metaData.schemaName + "." + metaData.tableName + "." + metaData.columnName );
+            }
+
+            ArrayList<String[]> data = new ArrayList<>();
+            for ( List<Object> row : rows ) {
+                String[] temp = new String[row.size()];
+                int counter = 0;
+                for ( Object o : row ) {
```
Comment (schoenja) [inline core/src/main/java/ch/unibas/dmi/dbis/polyphenydb/statistic/StatisticQueryProcessor.java]: Maybe consider using a `for (int i = 0;...)` loop here.

File: core/src/main/java/ch/unibas/dmi/dbis/polyphenydb/statistic/StatisticsManager.java [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+
+
+    /**
+     * Reset all statistics and reevaluate them
+     */
+    private void reevaluateStore() {
+        if ( this.sqlQueryInterface == null ) {
+            return;
+        }
+        log.warn( "Resetting StatisticStore." );
```
Comment (schoenja) [inline core/src/main/java/ch/unibas/dmi/dbis/polyphenydb/statistic/StatisticsManager.java]: Is `warn` the proper log level here? Or could it be lower? (`info` or even `debug`)

(Also for the second log statement in this method)

File: core/src/main/java/ch/unibas/dmi/dbis/polyphenydb/statistic/StatisticsManager.java [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+        String tableName = StatisticQueryProcessor.buildName( column.getSchema(), column.getTable() );
+        String columnName = StatisticQueryProcessor.buildName( column.getSchema(), column.getTable(), column.getName() );
+        String query = "SELECT COUNT(" + columnName + ") FROM " + tableName;
+        StatisticQueryColumn res = this.sqlQueryInterface.selectOneStat( query );
+
+        if ( res != null && res.getData() != null && res.getData().length != 0 ) {
+            try {
+                return Integer.parseInt( res.getData()[0] );
+            } catch ( NumberFormatException e ) {
+                log.error( "Count could not be parsed." );
```
Comment (schoenja) [inline core/src/main/java/ch/unibas/dmi/dbis/polyphenydb/statistic/StatisticsManager.java]: Please also include the Exception as well as the column name in the logging statement.
```suggestion
                log.error( "Count could not be parsed for column {}.", column.getQualifiedColumnName() e );
```

File: core/src/main/java/ch/unibas/dmi/dbis/polyphenydb/statistic/model/LimitedOccurrenceMap.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+import java.util.TreeMap;
+import lombok.extern.slf4j.Slf4j;
+
+
+/**
+ * Special implementation of a TreeMap to only allow a regulated size of entries
+ */
+@Slf4j
+public
+class LimitedOccurrenceMap<K> {
```
Comment (schoenja) [inline core/src/main/java/ch/unibas/dmi/dbis/polyphenydb/statistic/model/LimitedOccurrenceMap.java]: Please put this on one line.

File: dbms/src/main/java/ch/unibas/dmi/dbis/polyphenydb/PolyphenyDb.java [hunk 5]
```diff
@@ -156,20 +158,28 @@ public void join( final long millis ) throws InterruptedException {
         final Authenticator authenticator = new AuthenticatorImpl();
         final JdbcInterface jdbcInterface = new JdbcInterface( transactionManager, authenticator );
         final HttpServer httpServer = new HttpServer( transactionManager, authenticator, RuntimeConfig.WEBUI_SERVER_PORT.getInteger() );
+        // prolly overkill just for testing
+        final StatisticQueryProcessor statisticQueryProcessor = new StatisticQueryProcessor( transactionManager, authenticator );
```
Comment (schoenja) [inline dbms/src/main/java/ch/unibas/dmi/dbis/polyphenydb/PolyphenyDb.java]: Is this required for the actual module or only for testing? Can it be removed? Also in what way is it "overkill"?

File: core/src/main/java/ch/unibas/dmi/dbis/polyphenydb/statistic/StatisticQueryProcessor.java [hunk 6]
```diff
@@ -103,7 +103,7 @@ private String queryLimitCheck( String query ) {
     /**
      * Method to get all schemas, tables, and their columns in a database
      */
-    public ArrayList<ArrayList<String>> getSchemaTree() {
+    public List<List<String>> getSchemaTree() {
 
         ArrayList<ArrayList<String>> result = new ArrayList<>();
```
Comment (schoenja) [inline core/src/main/java/ch/unibas/dmi/dbis/polyphenydb/statistic/StatisticQueryProcessor.java]: My bad, I forgot to mention that also the type signatures in the method itself have to be adjusted.

**Comment by datomo** [comment]
Thanks @vogti and @schoenja for your corrections, suggestion and feedback.
After working through them, I kindly ask for a second audit if possible.

**Comment by schoenja** [comment]
Thanks for all the additional work. I will try to have a review up by tomorrow evening, otherwise it'll be not before monday.

@vogti if that is too late for you feel free to ignore me for the second review.

File: statistic/src/main/java/ch/unibas/dmi/dbis/polyphenydb/statistic/StatisticQueryProcessor.java [hunk 7]
```diff
@@ -0,0 +1,452 @@
...
+
+                    List<String> table = new ArrayList<>();
+
+                    for ( CatalogColumn catalogColumn : combinedTable.getColumns() ) {
+                        table.add( combinedSchema.getSchema().name + "." + combinedTable.getTable().name + "." + catalogColumn.name );
+                    }
+
+                    if ( combinedTable.getTable().tableType.equals( "TABLE" ) ) {
+                        tables.addAll( table );
+                    }
```
Comment (vogti) [inline statistic/src/main/java/ch/unibas/dmi/dbis/polyphenydb/statistic/StatisticQueryProcessor.java]: I'm not sure if this is working. You are comparing an Enum with a string. Correct would be
```java
if ( combinedTable.getTable().tableType == TableType.TABLE ) {
    tables.addAll( table );
}
```

**Comment by datomo** [comment]
Thanks for the work and feedback @vogti , I addressed your desired changes.
Let me know if there are still things to alter.

File: core/build.gradle [hunk 8]
```diff
@@ -45,9 +45,9 @@ apply plugin: "idea"
 apply plugin: "io.freefair.lombok"
 
 
-tasks.withType(JavaCompile) {
-    options.encoding = "UTF-8"
-}
+compileJava.options.encoding = 'UTF-8'
+compileTestJava.options.encoding = 'UTF-8'
+javadoc.options.encoding = 'UTF-8'
```
[the remark this answers is not in the document]
Comment (vogti) [reply core/build.gradle]: Yes, I will change this but not as part of this PR.

[the remark this answers is not in the document]
Comment (schoenja) [reply core/build.gradle]: Moved to new issue #196

**Pull request merged**